{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Within-host SNVs in individuals with multiple SCAN samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_df(file):\n",
    "    '''\n",
    "    Loads metadata tsv as df.\n",
    "    '''\n",
    "    with open(file) as tfile:\n",
    "        df = pd.read_csv(tfile, sep = '\\t')\n",
    "        if 'nwgc_id' in df.columns:\n",
    "            df['nwgc_id'] = df.nwgc_id.astype('str')\n",
    "        if 'position' and 'variant' in df.columns:\n",
    "            df['snv'] = df['position'].astype('str').str.cat(df['variant'],sep=\"\")\n",
    "        df.sort_values(by=['individual_identifier', 'date'], inplace=True)\n",
    "    return df\n",
    "\n",
    "metadata = load_df('../results/metadata_scan_all.tsv')\n",
    "\n",
    "snvs = load_df('../results/snvs_scan_all.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_genomes(file):\n",
    "    '''\n",
    "    Loads genomes as dictionary\n",
    "    '''\n",
    "    with open(file, 'r') as f:\n",
    "        genomes = SeqIO.to_dict(SeqIO.parse(f, 'fasta'))\n",
    "    return genomes\n",
    "\n",
    "genomes = load_genomes('../results/aligned.fasta')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find individuals with multiple samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_multiples(df):\n",
    "    '''\n",
    "    Returns df with just samples from multiple individuals\n",
    "    '''\n",
    "    multiples = df.groupby(by=['individual_identifier']).filter(lambda x: len(x) > 1).reset_index(drop=True)\n",
    "    n_individ = int(len(multiples)/2)\n",
    "    multiples['individual'] = [i for num in range(1, n_individ + 1) for i in [num]*2]\n",
    "    return multiples\n",
    "\n",
    "multiple_df = get_multiples(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>nwgc_id</th>\n",
       "      <th>strain</th>\n",
       "      <th>avg_ct</th>\n",
       "      <th>date</th>\n",
       "      <th>location</th>\n",
       "      <th>individual_identifier</th>\n",
       "      <th>age</th>\n",
       "      <th>puma</th>\n",
       "      <th>address_identifier</th>\n",
       "      <th>symptom_onset</th>\n",
       "      <th>individual</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20200416_fastq</td>\n",
       "      <td>388977</td>\n",
       "      <td>USA/WA-S301/2020</td>\n",
       "      <td>23.063411</td>\n",
       "      <td>2020-03-14</td>\n",
       "      <td>Pierce County</td>\n",
       "      <td>3f8fb75bfe48f31dc142c5bd629fc8d542f9a89f4973bd...</td>\n",
       "      <td>17.00</td>\n",
       "      <td>5311503.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20200416_fastq</td>\n",
       "      <td>389358</td>\n",
       "      <td>USA/WA-S286/2020</td>\n",
       "      <td>29.704358</td>\n",
       "      <td>2020-03-15</td>\n",
       "      <td>Pierce County</td>\n",
       "      <td>3f8fb75bfe48f31dc142c5bd629fc8d542f9a89f4973bd...</td>\n",
       "      <td>17.00</td>\n",
       "      <td>5311503.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20200831_fastq</td>\n",
       "      <td>438095</td>\n",
       "      <td>USA/WA-S2719/2020</td>\n",
       "      <td>21.992674</td>\n",
       "      <td>2020-07-02</td>\n",
       "      <td>King County</td>\n",
       "      <td>628b6616a28049b1099959c884f083a3be62e8983ba7a8...</td>\n",
       "      <td>85.00</td>\n",
       "      <td>5311613.0</td>\n",
       "      <td>b4b7055fc0b861638213a93cff8d5898b8990f18a536cb...</td>\n",
       "      <td>2020-07-01</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20200909_fastq</td>\n",
       "      <td>444373</td>\n",
       "      <td>USA/WA-S2769/2020</td>\n",
       "      <td>26.053100</td>\n",
       "      <td>2020-07-13</td>\n",
       "      <td>King County</td>\n",
       "      <td>628b6616a28049b1099959c884f083a3be62e8983ba7a8...</td>\n",
       "      <td>85.00</td>\n",
       "      <td>5311613.0</td>\n",
       "      <td>b4b7055fc0b861638213a93cff8d5898b8990f18a536cb...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20200909_fastq</td>\n",
       "      <td>457563</td>\n",
       "      <td>USA/WA-S2771/2020</td>\n",
       "      <td>23.799139</td>\n",
       "      <td>2020-08-01</td>\n",
       "      <td>King County</td>\n",
       "      <td>d9149c24ce3b5bffce436c93616efce1c66848d794e1a8...</td>\n",
       "      <td>63.75</td>\n",
       "      <td>5311607.0</td>\n",
       "      <td>fb859bab379819770195f951ddf2ba45d3402656ca8eab...</td>\n",
       "      <td>2020-07-26</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>20200909_fastq</td>\n",
       "      <td>472208</td>\n",
       "      <td>USA/WA-S2788/2020</td>\n",
       "      <td>28.146982</td>\n",
       "      <td>2020-08-12</td>\n",
       "      <td>King County</td>\n",
       "      <td>d9149c24ce3b5bffce436c93616efce1c66848d794e1a8...</td>\n",
       "      <td>63.83</td>\n",
       "      <td>5311607.0</td>\n",
       "      <td>fb859bab379819770195f951ddf2ba45d3402656ca8eab...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            batch nwgc_id             strain     avg_ct        date  \\\n",
       "0  20200416_fastq  388977   USA/WA-S301/2020  23.063411  2020-03-14   \n",
       "1  20200416_fastq  389358   USA/WA-S286/2020  29.704358  2020-03-15   \n",
       "2  20200831_fastq  438095  USA/WA-S2719/2020  21.992674  2020-07-02   \n",
       "3  20200909_fastq  444373  USA/WA-S2769/2020  26.053100  2020-07-13   \n",
       "4  20200909_fastq  457563  USA/WA-S2771/2020  23.799139  2020-08-01   \n",
       "5  20200909_fastq  472208  USA/WA-S2788/2020  28.146982  2020-08-12   \n",
       "\n",
       "        location                              individual_identifier    age  \\\n",
       "0  Pierce County  3f8fb75bfe48f31dc142c5bd629fc8d542f9a89f4973bd...  17.00   \n",
       "1  Pierce County  3f8fb75bfe48f31dc142c5bd629fc8d542f9a89f4973bd...  17.00   \n",
       "2    King County  628b6616a28049b1099959c884f083a3be62e8983ba7a8...  85.00   \n",
       "3    King County  628b6616a28049b1099959c884f083a3be62e8983ba7a8...  85.00   \n",
       "4    King County  d9149c24ce3b5bffce436c93616efce1c66848d794e1a8...  63.75   \n",
       "5    King County  d9149c24ce3b5bffce436c93616efce1c66848d794e1a8...  63.83   \n",
       "\n",
       "        puma                                 address_identifier symptom_onset  \\\n",
       "0  5311503.0                                                NaN           NaN   \n",
       "1  5311503.0                                                NaN           NaN   \n",
       "2  5311613.0  b4b7055fc0b861638213a93cff8d5898b8990f18a536cb...    2020-07-01   \n",
       "3  5311613.0  b4b7055fc0b861638213a93cff8d5898b8990f18a536cb...           NaN   \n",
       "4  5311607.0  fb859bab379819770195f951ddf2ba45d3402656ca8eab...    2020-07-26   \n",
       "5  5311607.0  fb859bab379819770195f951ddf2ba45d3402656ca8eab...           NaN   \n",
       "\n",
       "   individual  \n",
       "0           1  \n",
       "1           1  \n",
       "2           2  \n",
       "3           2  \n",
       "4           3  \n",
       "5           3  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiple_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So three individuals have multiple samples. In one person, samples are separated by 1 day. In two people, samples are separated by 11 days. All persons have a \"low (<24) Ct\" sample and a higher Ct sample. I'm not convinced that variants can be reliably called from high Ct samples, but this might be an unavoidable challenge of having samples from across an infection time course.\n",
    "\n",
    "\n",
    "## Comparing consensus genomes from each person's samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: {'location': [array([], dtype=int64)], 'base1': [], 'base2': []},\n",
       " 2: {'location': [array([27689])], 'base1': ['T'], 'base2': ['C']},\n",
       " 3: {'location': [array([], dtype=int64)], 'base1': [], 'base2': []}}"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def convert(df, genomes):\n",
    "    '''\n",
    "    Checks if consensus genomes in samples from same people are identical to each other.\n",
    "    '''\n",
    "    mapping = {}\n",
    "    for person in df.individual.unique():\n",
    "        strains = df.loc[df.individual==person, 'strain']\n",
    "        for strain in strains:\n",
    "            seq = list(genomes[strain].seq)\n",
    "            seq_conversion = {\n",
    "                    'C': 2,\n",
    "                    'T': 3,\n",
    "                    'G': 5,\n",
    "                    'A': 7,\n",
    "                    'D': 105,\n",
    "                    'R': 35,\n",
    "                    'Y': 6,\n",
    "                    'H': 42,\n",
    "                    'K': 15,\n",
    "                    'V': 70,\n",
    "                    'N': 210,\n",
    "                    'M': 14,\n",
    "                    'W': 21,\n",
    "                    'S': 10,\n",
    "                    '-': 210\n",
    "            }\n",
    "            sequence = [seq_conversion.get(base) for base in seq]\n",
    "            array = np.asarray(sequence, dtype=int)\n",
    "            mapping[strain] = array\n",
    "    return mapping\n",
    "\n",
    "multiple_genomes = convert(multiples_df, genomes)\n",
    "\n",
    "def compare(array1, array2):\n",
    "    '''\n",
    "    Identifies SNPs between sequences where sequences are represented as numpy integer arrays.\n",
    "    CTGA must be represented by prime numbers and ambiguous bases are the multiples of those prime numbers.\n",
    "    '''\n",
    "    comparison = {}\n",
    "    condition1 = (array1 != array2)\n",
    "    condition2 = (array1 % array2 != 0)\n",
    "    condition3 = (array2 % array1 != 0)\n",
    "    location = np.where(condition1 & condition2 & condition3)\n",
    "    position = [pos + 1 for pos in location] # Plus one makes sure that bp numbering begins at 1, not zero\n",
    "    base1 = array1[location]\n",
    "    base2 = array2[location]\n",
    "    seq_conversion = {\n",
    "        2:'C',\n",
    "        3:'T',\n",
    "        5:'G',\n",
    "        7:'A',\n",
    "        105:'D',\n",
    "        35:'R',\n",
    "        6:'Y',\n",
    "        42:'H',\n",
    "        15:'K',\n",
    "        70:'V',\n",
    "        210:'N',\n",
    "        14:'M',\n",
    "        21:'W',\n",
    "        10:'S'\n",
    "    }\n",
    "    comparison['location'] = position\n",
    "    comparison['base1'] = [seq_conversion.get(base) for base in base1]\n",
    "    comparison['base2'] = [seq_conversion.get(base) for base in base2]\n",
    "    return comparison\n",
    "\n",
    "def compare_pairwise(df, mapping):\n",
    "    '''\n",
    "    Compares all strains pairwise storing the genetic distance between each strains as NxN arrays at the genome & segment level.\n",
    "    '''\n",
    "    pairwise = {}\n",
    "    for person in df.individual.unique():\n",
    "        strains = df.loc[df.individual==person, 'strain'].reset_index(drop=True)\n",
    "        pairwise[person] = compare(mapping[strains[0]], mapping[strains[1]])\n",
    "    return pairwise\n",
    "\n",
    "compare_pairwise(multiple_df, multiple_genomes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So consensus genomes are identical for persons 1 & 3. Consensus genome has one substitution for person 2: T27689C.\n",
    "\n",
    "\n",
    "## Comparing iSNVs from each person's samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snvs(df, multiples):\n",
    "    '''\n",
    "    Filters snvs dataframe to individuals with multiple samples\n",
    "    '''\n",
    "    ids = multiples['nwgc_id']\n",
    "    filtered = df[df.nwgc_id.isin(ids)].reset_index(drop=True)\n",
    "    merged = filtered.merge(multiples, how='left')\n",
    "    return merged\n",
    "\n",
    "multiple_snvs = filter_snvs(snvs, multiples_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Person: 1\n",
      " # of iSNVs in sample: 22\n",
      " # of iSNVs in sample: 58\n",
      "Person: 2\n",
      " # of iSNVs in sample: 1\n",
      " # of iSNVs in sample: 69\n",
      "Person: 3\n",
      " # of iSNVs in sample: 1\n",
      " # of iSNVs in sample: 22\n"
     ]
    }
   ],
   "source": [
    "def n_snvs(variants, metadata):\n",
    "    for person in metadata.individual.unique():\n",
    "        print('Person: ' + str(person))\n",
    "        for strain in metadata.loc[metadata.individual==person, 'nwgc_id']:\n",
    "            n_samples = len(variants[variants.nwgc_id==strain])\n",
    "            print(' # of iSNVs in sample: ' + str(n_samples))\n",
    "\n",
    "n_snvs(multiple_snvs, multiple_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that all the highr Ct (later timepoint) samples have more variants. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>nwgc_id</th>\n",
       "      <th>position</th>\n",
       "      <th>variant</th>\n",
       "      <th>genome</th>\n",
       "      <th>protein</th>\n",
       "      <th>aa_position</th>\n",
       "      <th>aa_variant</th>\n",
       "      <th>aa_genome</th>\n",
       "      <th>mut_type</th>\n",
       "      <th>...</th>\n",
       "      <th>frequency</th>\n",
       "      <th>symptom_onset</th>\n",
       "      <th>batch</th>\n",
       "      <th>age</th>\n",
       "      <th>date</th>\n",
       "      <th>avg_ct</th>\n",
       "      <th>address_identifier</th>\n",
       "      <th>snv</th>\n",
       "      <th>individual</th>\n",
       "      <th>is_shared</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>USA/WA-S301/2020</td>\n",
       "      <td>388977</td>\n",
       "      <td>9502</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ORF1a</td>\n",
       "      <td>3079</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>0.058146</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20200416_fastq</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2020-03-14</td>\n",
       "      <td>23.063411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9502T</td>\n",
       "      <td>1</td>\n",
       "      <td>shared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>USA/WA-S301/2020</td>\n",
       "      <td>388977</td>\n",
       "      <td>22133</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>191</td>\n",
       "      <td>K</td>\n",
       "      <td>E</td>\n",
       "      <td>NS</td>\n",
       "      <td>...</td>\n",
       "      <td>0.013393</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20200416_fastq</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2020-03-14</td>\n",
       "      <td>23.063411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22133A</td>\n",
       "      <td>1</td>\n",
       "      <td>shared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>USA/WA-S286/2020</td>\n",
       "      <td>389358</td>\n",
       "      <td>9502</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ORF1a</td>\n",
       "      <td>3079</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>0.170732</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20200416_fastq</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2020-03-15</td>\n",
       "      <td>29.704358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9502T</td>\n",
       "      <td>1</td>\n",
       "      <td>shared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>USA/WA-S286/2020</td>\n",
       "      <td>389358</td>\n",
       "      <td>22133</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>191</td>\n",
       "      <td>K</td>\n",
       "      <td>E</td>\n",
       "      <td>NS</td>\n",
       "      <td>...</td>\n",
       "      <td>0.033184</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20200416_fastq</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2020-03-15</td>\n",
       "      <td>29.704358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22133A</td>\n",
       "      <td>1</td>\n",
       "      <td>shared</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              strain nwgc_id  position variant genome protein aa_position  \\\n",
       "11  USA/WA-S301/2020  388977      9502       T      C   ORF1a        3079   \n",
       "18  USA/WA-S301/2020  388977     22133       A      G       S         191   \n",
       "30  USA/WA-S286/2020  389358      9502       T      C   ORF1a        3079   \n",
       "57  USA/WA-S286/2020  389358     22133       A      G       S         191   \n",
       "\n",
       "   aa_variant aa_genome mut_type  ... frequency symptom_onset           batch  \\\n",
       "11          A         A        S  ...  0.058146           NaN  20200416_fastq   \n",
       "18          K         E       NS  ...  0.013393           NaN  20200416_fastq   \n",
       "30          A         A        S  ...  0.170732           NaN  20200416_fastq   \n",
       "57          K         E       NS  ...  0.033184           NaN  20200416_fastq   \n",
       "\n",
       "     age        date     avg_ct  address_identifier     snv  individual  \\\n",
       "11  17.0  2020-03-14  23.063411                 NaN   9502T           1   \n",
       "18  17.0  2020-03-14  23.063411                 NaN  22133A           1   \n",
       "30  17.0  2020-03-15  29.704358                 NaN   9502T           1   \n",
       "57  17.0  2020-03-15  29.704358                 NaN  22133A           1   \n",
       "\n",
       "   is_shared  \n",
       "11    shared  \n",
       "18    shared  \n",
       "30    shared  \n",
       "57    shared  \n",
       "\n",
       "[4 rows x 23 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def compare_snvs(df):\n",
    "    '''\n",
    "    Adds whether or not a SNV in one sample is shared with the other sample from that person.\n",
    "    '''\n",
    "    counter = 0\n",
    "    for person in df.individual.unique():\n",
    "        person_df = df[df.individual==person].copy()       \n",
    "        shared_snvs = person_df.groupby('snv').filter(lambda x: len(x) > 1)\n",
    "        snvs = shared_snvs['snv'].unique()\n",
    "        status = []\n",
    "        for i in person_df.index:\n",
    "            if person_df.loc[i, 'snv'] in snvs:\n",
    "                status.append('shared')\n",
    "            else:\n",
    "                status.append('not_shared')\n",
    "        person_df['is_shared'] = status\n",
    "        if counter == 0:\n",
    "            joined_df = person_df\n",
    "            counter = 1\n",
    "        else:\n",
    "            joined_df = joined_df.append(person_df)\n",
    "    return joined_df\n",
    "\n",
    "multiple_snvs = compare_snvs(multiple_snvs)\n",
    "\n",
    "multiple_snvs[multiple_snvs.is_shared=='shared']\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Only person 1 has shared iSNVs: 9502T & 22133A. Frequency of both variants increases in second sample compared to the first sample. These are iSNVs that are relatively common in the SCAN dataset. I'm somewhat surprised that more iSNVs aren't shared considering the samples are only one day apart; it boosts my suspicion that we're having a hard time reliably calling variants in high Ct samples (especially w/o replicate sequencing to verify variants)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>nwgc_id</th>\n",
       "      <th>position</th>\n",
       "      <th>variant</th>\n",
       "      <th>genome</th>\n",
       "      <th>protein</th>\n",
       "      <th>aa_position</th>\n",
       "      <th>aa_variant</th>\n",
       "      <th>aa_genome</th>\n",
       "      <th>mut_type</th>\n",
       "      <th>...</th>\n",
       "      <th>frequency</th>\n",
       "      <th>symptom_onset</th>\n",
       "      <th>batch</th>\n",
       "      <th>age</th>\n",
       "      <th>date</th>\n",
       "      <th>avg_ct</th>\n",
       "      <th>address_identifier</th>\n",
       "      <th>snv</th>\n",
       "      <th>individual</th>\n",
       "      <th>is_shared</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>USA/WA-S2719/2020</td>\n",
       "      <td>438095</td>\n",
       "      <td>27689</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ORF7a</td>\n",
       "      <td>99</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>NS</td>\n",
       "      <td>...</td>\n",
       "      <td>0.066038</td>\n",
       "      <td>2020-07-01</td>\n",
       "      <td>20200831_fastq</td>\n",
       "      <td>85.0</td>\n",
       "      <td>2020-07-02</td>\n",
       "      <td>21.992674</td>\n",
       "      <td>b4b7055fc0b861638213a93cff8d5898b8990f18a536cb...</td>\n",
       "      <td>27689C</td>\n",
       "      <td>2</td>\n",
       "      <td>not_shared</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               strain nwgc_id  position variant genome protein aa_position  \\\n",
       "80  USA/WA-S2719/2020  438095     27689       C      T   ORF7a          99   \n",
       "\n",
       "   aa_variant aa_genome mut_type  ... frequency symptom_onset           batch  \\\n",
       "80          P         L       NS  ...  0.066038    2020-07-01  20200831_fastq   \n",
       "\n",
       "     age        date     avg_ct  \\\n",
       "80  85.0  2020-07-02  21.992674   \n",
       "\n",
       "                                   address_identifier     snv  individual  \\\n",
       "80  b4b7055fc0b861638213a93cff8d5898b8990f18a536cb...  27689C           2   \n",
       "\n",
       "     is_shared  \n",
       "80  not_shared  \n",
       "\n",
       "[1 rows x 23 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiple_snvs[multiple_snvs.nwgc_id=='438095']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The one iSNV in Person 2's first sample is also the same variant that was fixed in the consensus genome of the second sample. That's cool!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nextstrain",
   "language": "python",
   "name": "nextstrain"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
